relay-server/src/actors/project_cache.rs

```diff
@@ -400,6 +374,109 @@ struct UpdateProjectState {
     no_cache: bool,
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+struct QueueKey {
+    key: ProjectKey,
+    sampling_key: ProjectKey,
+}
+
+/// The queue (buffer) of the incoming envelopes.
+#[derive(Debug, Default)]
+struct Queue {
+    /// Contains the cache of the incoming envelopes.
+    buffer: HashMap<QueueKey, Vec<(Box<Envelope>, EnvelopeContext)>>,
+    /// Index of the buffered project keys.
+    index: HashMap<ProjectKey, HashSet<QueueKey>>,
```

Comment:
nit 1: We tend to use `BTreeSet` for small sets, because the `log(N)` lookup might be faster than hashing and comparing keys when the tree is small.
```suggestion
    index: HashMap<ProjectKey, BTreeSet<QueueKey>>,
```

See also https://github.com/Rufflewind/bench-maps.

Reply:
makes sense , done in [435ac7b](https://github.com/getsentry/relay/pull/1879/commits/435ac7b6203eb7fc8ff74f15f05c68f043767551)